dali/pipeline/operator/sequence_shape.h

```diff
@@ -0,0 +1,201 @@
+TensorVector<Backend> unfold_outer_dims(const TensorVector<Backend> &data, int num_unfold_dims) {
+  const auto &shape = data.shape();
+  auto group_shapes = shape.first(num_unfold_dims);
+  const auto num_unfolded_samples = group_shapes.num_elements();
+  TensorVector<Backend> unfolded_tensor(num_unfolded_samples);
+  auto type_info = data.type_info();
+  auto type = type_info.id();
+  auto is_pinned = data.is_pinned();
+  auto order = data.order();
+  unfolded_tensor.set_type(type);
+  unfolded_tensor.set_pinned(is_pinned);
+  unfolded_tensor.set_order(order);
+  int elem_idx = 0;
+  for (int sample_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    const auto &sample_shape = shape[sample_idx];
+    int num_frames = volume(sample_shape.begin(), sample_shape.begin() + num_unfold_dims);
+    TensorShape<> element_shape{sample_shape.begin() + num_unfold_dims, sample_shape.end()};
+    int element_volume = volume(element_shape);
+    auto num_bytes = type_info.size() * element_volume;
+    uint8_t *base_ptr =
+        const_cast<uint8_t *>(static_cast<const uint8_t *>(data.raw_tensor(sample_idx)));
+    for (int frame_idx = 0; frame_idx < num_frames; ++frame_idx) {
+      uint8_t *ptr = base_ptr + frame_idx * num_bytes;
+      unfolded_tensor[elem_idx++].ShareData(ptr, num_bytes, is_pinned, element_shape, type, order);
+    }
+  }
+  assert(elem_idx == num_unfolded_samples);
+  return unfolded_tensor;
+}
+
+template <typename Backend>
+TensorList<Backend> unfold_outer_dims(const TensorList<Backend> &data, int num_unfold_dims) {
+  // TODO(ktokarski) TODO(klecki)
+  // Rework it when TensorList stops being contigious and supports "true sample" mode
+  const auto &shape = data.shape();
+  TensorList<Backend> tl;
+  tl.ShareData(data);
+  auto expanded_shape = unfold_outer_dims(shape, num_unfold_dims);
+  tl.Resize(expanded_shape, data.type());
+  return tl;
+}
+
+inline TensorListShape<> unfold_outer_dims(const TensorListShape<> &shape, int num_unfold_dims) {
+  if (num_unfold_dims == 0) {
+    return shape;
+  } else if (num_unfold_dims == 1) {
+    return unfold_outer_dim(shape);
+  } else {
+    auto data_shape = collapse_dims(shape, {{0, num_unfold_dims}});
+    return unfold_outer_dim(data_shape);
+  }
+}
+
+inline TensorListShape<> fold_outermost_like(const TensorListShape<> &shape,
+                                             const TensorListShape<> &folded_shape) {
+  if (folded_shape.sample_dim() == 0) {
+    return shape;
+  }
+  auto num_samples = folded_shape.num_samples();
+  TensorListShape<> res(num_samples, folded_shape.sample_dim() + shape.sample_dim());
+  for (int sample_idx = 0, element_idx = 0; sample_idx < num_samples; sample_idx++) {
+    auto group_shape = folded_shape[sample_idx];
+    auto num_elements = volume(group_shape);
+    DALI_ENFORCE(num_elements > 0,
+                 make_string("Zero-volume samples are not allowed. Got volume 0 for sample ",
+                             sample_idx, "."));
```

Comment:
This is a very strong assumption and one that's not necessary(*). Also, I think the error message is slightly wrong - it refers not to 0-volume samples but rather to 0-frame samples.

(*) it will be a problem if there are only 0-frame samples there.

Reply:
Right, the error message is slightly confusing, it is about 0-frame samples. It wasn't clear to me what should be done if such case is encountered: what should be a shape of a frame in that case? Should the sample be completely skipped? Wouldn't that lead to unexpected errors down the pipeline? I decided to stay on a explicit site and just forbid that. In any case the specific operator could override that coallescing if there is easier way to infer the output shape (like Guassian blur does where the output shape is the same as the input shape).

Reply:
I'd say we should skip such samples entirely - no frames, nothing to do. As for the remaining extents, there's one idea which would work with argument broadcasting (but would be much harder with per-frame arguments): run SetupImpl on a dummy frame (take the size from the non-expanded dimensions of the empty sequence) and use that size for the frame shape. Unfortunately, per-frame arguments could mess with the process.